### PR TITLE
Cache Docker Pulls as SIFs

### LINKED ISSUE #2739: Pull should cache OCI Images as SIFs
## Version of Singularity:

3.1.0-rc3

### Expected behavior

Repeated pulls should copy a cached SIF if the image manifest doesn't change. This is how `actions` commands are currently working.

### Actual behavior

Always builds an new SIF from layers.

### Steps to reproduce behavior

```
$ singularity pull c.sif docker://alpine:latest
WARNING: Authentication token file not found : Only pulls of public images will succeed
INFO:    Starting build...
Getting image source signatures
Skipping fetch of repeat blob sha256:6c40cc604d8e4c121adcb6b0bfe8bb038815c350980090e74aa5a6423f8f82c0
Copying config sha256:57c41aa0b96eb7fcf351321550126306e1da1bc1b1c1c847f4976fd201d95ba1
 585 B / 585 B [============================================================] 0s
Writing manifest to image destination
Storing signatures
INFO:    Creating SIF file...
INFO:    Build complete: c.sif
$ singularity pull c.sif docker://alpine:latest
WARNING: Authentication token file not found : Only pulls of public images will succeed
INFO:    Starting build...
Getting image source signatures
Skipping fetch of repeat blob sha256:6c40cc604d8e4c121adcb6b0bfe8bb038815c350980090e74aa5a6423f8f82c0
Copying config sha256:57c41aa0b96eb7fcf351321550126306e1da1bc1b1c1c847f4976fd201d95ba1
 585 B / 585 B [============================================================] 0s
Writing manifest to image destination
Storing signatures
INFO:    Creating SIF file...
INFO:    Build complete: c.sif
```

### PR DESCRIPTION
**Pull request by WestleyK**
## Description of the Pull Request (PR):

This PR will cache container pulls from docker as SIFs.

## This fixes or addresses the following GitHub issues:

- Fixes #2739 

<br>

### CODE REVIEW

File: cmd/internal/cli/pull.go [hunk 1]
```diff
@@ -228,13 +232,69 @@ func pullRun(cmd *cobra.Command, args []string) {
...
 		}
 
-		libexec.PullOciImage(name, args[i], types.Options{
-			TmpDir:           tmpDir,
-			Force:            force,
-			NoHTTPS:          noHTTPS,
-			DockerAuthConfig: authConf,
-			NoCleanUp:        noCleanUp,
-		})
+		sysCtx := &ocitypes.SystemContext{
```
Comment (mem) [inline cmd/internal/cli/pull.go]: This function is already too long to be adding more lines to it :-|

Can you please move the whole ociclient.IsSupported(transport) case to a separate function, please?

File: cmd/internal/cli/pull.go [hunk 2]
```diff
@@ -228,13 +232,69 @@ func pullRun(cmd *cobra.Command, args []string) {
...
+			OCIInsecureSkipTLSVerify:    noHTTPS,
+			DockerInsecureSkipTLSVerify: noHTTPS,
+			DockerAuthConfig:            authConf,
+		}
+
+		sum, err := ociclient.ImageSHA(args[i], sysCtx)
+		if err != nil {
+			sylog.Fatalf("Failed to get SHA of %v: %v", args[i], err)
```
Comment (mem) [inline cmd/internal/cli/pull.go]:
```suggestion
			sylog.Fatalf("Failed to get checksum for %s: %s", args[i], err)
```

File: cmd/internal/cli/pull.go [hunk 3]
```diff
@@ -228,13 +232,69 @@ func pullRun(cmd *cobra.Command, args []string) {
...
+		}
+
+		name := uri.GetName(args[i])
+		imgabs := cache.OciTempImage(sum, name)
```
Comment (mem) [inline cmd/internal/cli/pull.go]: `imagePath` maybe? Or `cachedImg`?

File: cmd/internal/cli/pull.go [hunk 4]
```diff
@@ -228,13 +232,69 @@ func pullRun(cmd *cobra.Command, args []string) {
...
+
+		exists, err := cache.OciTempExists(sum, name)
```
Comment (mem) [inline cmd/internal/cli/pull.go]: it looks weird to be asking for the absolute path first and then ask if the image exists? Maybe flip these two lines?

File: cmd/internal/cli/pull.go [hunk 5]
```diff
@@ -228,13 +232,69 @@ func pullRun(cmd *cobra.Command, args []string) {
...
+		if err != nil {
+			sylog.Fatalf("Unable to check if %v exists: %v", imgabs, err)
```
Comment (mem) [inline cmd/internal/cli/pull.go]: Is it really ok to show this path to the user? Shouldn't you use the name provided by the user? Either `args[i]` or `name`?
```suggestion
			sylog.Fatalf("Unable to check if %s exists: %s", imgabs, err)
```

File: cmd/internal/cli/pull.go [hunk 6]
```diff
@@ -228,13 +232,69 @@ func pullRun(cmd *cobra.Command, args []string) {
...
+		}
+		if !exists {
+			sylog.Infof("Converting OCI blobs to SIF format")
```
Comment (mem) [inline cmd/internal/cli/pull.go]: the body of this if should be a function of its own, returning an error

File: cmd/internal/cli/pull.go [hunk 7]
```diff
@@ -228,13 +232,69 @@ func pullRun(cmd *cobra.Command, args []string) {
...
+		}
+
+		// Perms are 777 *prior* to umask
+		dstFile, err := os.OpenFile(name, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0777)
+		if err != nil {
+			sylog.Fatalf("Unable to open file: %s: %v\n", name, err)
+		}
+		defer dstFile.Close()
+
+		srcFile, err := os.OpenFile(imgabs, os.O_RDONLY, 0444)
```
Comment (mem) [inline cmd/internal/cli/pull.go]: mask is not used when reading a file, this is the same as `os.Open(imgabs)`

File: cmd/internal/cli/pull.go [hunk 8]
```diff
@@ -228,13 +232,69 @@ func pullRun(cmd *cobra.Command, args []string) {
...
+				sylog.Fatalf("Unable to build: %v", err)
+			}
+		} else {
+			sylog.Infof("Using cached image")
+		}
+
+		// Perms are 777 *prior* to umask
+		dstFile, err := os.OpenFile(name, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0777)
+		if err != nil {
+			sylog.Fatalf("Unable to open file: %s: %v\n", name, err)
```
Comment (mem) [inline cmd/internal/cli/pull.go]:
```suggestion
			sylog.Fatalf("Unable to open file for writing: %s: %v\n", name, err)
```

File: cmd/internal/cli/pull.go [hunk 9]
```diff
@@ -228,13 +232,69 @@ func pullRun(cmd *cobra.Command, args []string) {
...
+		}
+		defer dstFile.Close()
+
+		srcFile, err := os.OpenFile(imgabs, os.O_RDONLY, 0444)
+		if err != nil {
+			sylog.Fatalf("Unable to open file: %s: %v\n", imgabs, err)
```
Comment (mem) [inline cmd/internal/cli/pull.go]:
```suggestion
			sylog.Fatalf("Unable to open file for reading: %s: %v\n", imgabs, err)
```

**Review by mem — DISMISSED** [review]

File: cmd/internal/cli/pull.go (hunk 3 above)
Comment (WestleyK) [reply cmd/internal/cli/pull.go]: Yeah, will fix that. This was mostly copy-paste from `actions.go`.

File: cmd/internal/cli/pull.go (hunk 7 above)
Comment (WestleyK) [reply cmd/internal/cli/pull.go]: Can you explain a bit more?

**Review by ikaneshiro — APPROVED** [review]
Lgtm

**Pull request merged**